File: ros_beginner/src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!  /usr/bin/python3
import rclpy
from rclpy.node import Node
from functools import partial
from example_interfaces.srv import AddTwoInts


class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.get_logger().info("AddTwoIntsClient is established.")
        self.call_add_two_ints_server(10, 2)
        self.call_add_two_ints_server(4, 8)

    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for server...")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request=request)
        future.add_done_callback(partial(self.callback_future, request=request))

    def callback_future(self, future, request):
        try:
            response = future.result()
            self.get_logger().info(f"{request.a}+{request.b}={response.sum}")
        except Exception as e:
            self.get_logger().error(f"Error occured {e}")


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: ros_beginner/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

class AddTwoIntsClient : public rclcpp::Node {
   public:
    AddTwoIntsClient() : Node("add_two_ints_client") {
        RCLCPP_INFO(this->get_logger(), "AddTwoIntsClient is estabished");

        thread1_ = std::thread(
            std::bind(&AddTwoIntsClient::callAddTwoIntsServer, this, 10, 4));

        thread2_ = std::thread(
            std::bind(&AddTwoIntsClient::callAddTwoIntsServer, this, 3, 9));
    }

   private:
    void callAddTwoIntsServer(int a, int b) {

        auto client = this->create_client<example_interfaces::srv::AddTwoInts>(
            "add_two_ints");
        RCLCPP_INFO(this->get_logger(), "Into callback");
        while (!client->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_WARN(this->get_logger(),
                        "Waiting for the server to be up...");
        }

        RCLCPP_INFO(this->get_logger(), "Into callback");
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;
        auto future = client->async_send_request(request);

        try {

            auto response = future.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %ld", a, b,
                        response->sum);

        } catch (const std::exception& e) {
            RCLCPP_ERROR(this->get_logger(), "Error");
        }
    }

    std::thread thread1_;
    std::thread thread2_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClient>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}